File: other-programming-languages/Scala-002/2024/2024-016/2024-016.scala
import scala.collection.mutable
import scala.io.Source
import scala.collection.mutable.{ArrayDeque, Set}

object MazeSolver {
  // Directions: 0=East, 1=South, 2=West, 3=North
  val directions = Array(
    (0, 1),   // East
    (1, 0),   // South
    (0, -1),  // West
    (-1, 0)   // North
  )

  def parseInput(filename: String): Array[String] = {
    Source.fromFile(filename).getLines().toArray
  }

  def solveMaze(mazeLines: Array[String]): Option[Int] = {
    val rows = mazeLines.length
    val cols = mazeLines(0).length

    // Find S and E
    var start: Option[(Int, Int)] = None
    var end: Option[(Int, Int)] = None

    for (r <- 0 until rows; c <- 0 until cols) {
      mazeLines(r)(c) match {
        case 'S' => start = Some((r, c))
        case 'E' => end = Some((r, c))
        case _   =>
      }
    }

    if (start.isEmpty || end.isEmpty) {
      throw new IllegalArgumentException("Could not find 'S' or 'E' in the maze.")
    }

    val (startRow, startCol) = start.get
    val (endRow, endCol) = end.get

    // Dijkstra's algorithm
    val INF = Int.MaxValue
    val dist = Array.fill(rows, cols, 4)(INF)
    val startDir = 0 // Facing East
    dist(startRow)(startCol)(startDir) = 0

    val pq = mutable.PriorityQueue[(Int, Int, Int, Int)]()(Ordering.by(-_._1))
    pq.enqueue((0, startRow, startCol, startDir))

    val visited = mutable.Set[(Int, Int, Int)]()

    while (pq.nonEmpty) {
      val (cost, r, c, d) = pq.dequeue()

      // Skip already visited states
      if (visited.contains((r, c, d))) {
        // Skip to the next iteration
      } else {
        visited.add((r, c, d))

        // If we've reached the end, return the cost
        if ((r, c) == (endRow, endCol)) {
          return Some(cost)
        }

        // Explore neighbors:
        // 1) Move forward (cost + 1) if not a wall
        val (dr, dc) = directions(d)
        val nr = r + dr
        val nc = c + dc
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && mazeLines(nr)(nc) != '#') {
          val newCost = cost + 1
          if (newCost < dist(nr)(nc)(d)) {
            dist(nr)(nc)(d) = newCost
            pq.enqueue((newCost, nr, nc, d))
          }
        }

        // 2) Turn left (cost + 1000)
        val leftDir = (d + 3) % 4
        val leftCost = cost + 1000
        if (leftCost < dist(r)(c)(leftDir)) {
          dist(r)(c)(leftDir) = leftCost
          pq.enqueue((leftCost, r, c, leftDir))
        }

        // 3) Turn right (cost + 1000)
        val rightDir = (d + 1) % 4
        val rightCost = cost + 1000
        if (rightCost < dist(r)(c)(rightDir)) {
          dist(r)(c)(rightDir) = rightCost
          pq.enqueue((rightCost, r, c, rightDir))
        }
      }
    }

    None // If no path is found
  }

  def solvePart2(mazeLines: Array[String]): Int = {
    val rows = mazeLines.length
    val cols = mazeLines(0).length

    var start: Option[(Int, Int)] = None
    var end: Option[(Int, Int)] = None

    // Identify 'S' and 'E'
    for (r <- 0 until rows; c <- 0 until cols) {
      mazeLines(r)(c) match {
        case 'S' => start = Some((r, c))
        case 'E' => end = Some((r, c))
        case _   =>
      }
    }

    if (start.isEmpty || end.isEmpty) {
      throw new IllegalArgumentException("Could not find 'S' or 'E' in the maze.")
    }

    val (startRow, startCol) = start.get
    val (endRow, endCol) = end.get

    val INF = Int.MaxValue
    val dist = Array.fill(rows, cols, 4)(INF)
    val startDir = 0 // Facing East
    dist(startRow)(startCol)(startDir) = 0

    val pq = mutable.PriorityQueue[(Int, Int, Int, Int)]()(Ordering.by(-_._1))
    pq.enqueue((0, startRow, startCol, startDir))

    val visited = mutable.Set[(Int, Int, Int)]()

    while (pq.nonEmpty) {
      val (cost, r, c, d) = pq.dequeue()

      if (visited.contains((r, c, d))) {
        // Skip already visited states
      } else {
        visited.add((r, c, d))

        // Move forward
        val (dr, dc) = directions(d)
        val nr = r + dr
        val nc = c + dc
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && mazeLines(nr)(nc) != '#') {
          val newCost = cost + 1
          if (newCost < dist(nr)(nc)(d)) {
            dist(nr)(nc)(d) = newCost
            pq.enqueue((newCost, nr, nc, d))
          }
        }

        // Turn left
        val leftDir = (d + 3) % 4
        val leftCost = cost + 1000
        if (leftCost < dist(r)(c)(leftDir)) {
          dist(r)(c)(leftDir) = leftCost
          pq.enqueue((leftCost, r, c, leftDir))
        }

        // Turn right
        val rightDir = (d + 1) % 4
        val rightCost = cost + 1000
        if (rightCost < dist(r)(c)(rightDir)) {
          dist(r)(c)(rightDir) = rightCost
          pq.enqueue((rightCost, r, c, rightDir))
        }
      }
    }

    // Find minimal cost to reach the end
    val minCostEnd = (0 until 4).map(d => dist(endRow)(endCol)(d)).min
    if (minCostEnd == INF) {
      return 0
    }

    // Mark cells on the best path
    val onBestPath = Array.fill(rows, cols)(false)

    val queue = ArrayDeque[(Int, Int, Int)]()
    for (d <- 0 until 4 if dist(endRow)(endCol)(d) == minCostEnd) {
      queue.append((endRow, endCol, d))
    }

    val visitedRev = mutable.Set[(Int, Int, Int)]()
    queue.foreach(visitedRev.add)

    while (queue.nonEmpty) {
      val (r, c, d) = queue.removeHead()
      onBestPath(r)(c) = true

      val costHere = dist(r)(c)(d)

      // Predecessor by moving forward
      val (dr, dc) = directions(d)
      val rPrev = r - dr
      val cPrev = c - dc
      if (rPrev >= 0 && rPrev < rows && cPrev >= 0 && cPrev < cols && mazeLines(rPrev)(cPrev) != '#') {
        if (dist(rPrev)(cPrev)(d) == costHere - 1 && !visitedRev.contains((rPrev, cPrev, d))) {
          visitedRev.add((rPrev, cPrev, d))
          queue.append((rPrev, cPrev, d))
        }
      }

      // Predecessor by turning
      for (dPrev <- Seq((d + 3) % 4, (d + 1) % 4)) {
        if (dist(r)(c)(dPrev) == costHere - 1000 && !visitedRev.contains((r, c, dPrev))) {
          visitedRev.add((r, c, dPrev))
          queue.append((r, c, dPrev))
        }
      }
    }

    // Count tiles on at least one best path
    onBestPath.flatten.count(_ == true)
  }

  def main(args: Array[String]): Unit = {
    val lines = parseInput("\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-016\\input.txt")
    solveMaze(lines) match {
      case Some(answer) => println(s"Lowest possible score: $answer")
      case None         => println("No path found!")
    }

    val result = solvePart2(lines)
    println(s"Number of tiles on at least one best path: $result")
  }
}

File: other-programming-languages/Rust-003/2024/2024-016/2024-016.rs
use std::collections::{BinaryHeap, HashSet, VecDeque};
use std::cmp::Reverse;

fn solve_maze(maze_lines: &[String]) -> Option<usize> {
    // Directions: 0=East, 1=South, 2=West, 3=North
    let directions = [
        (0, 1),   // East
        (1, 0),   // South
        (0, -1),  // West
        (-1, 0),  // North
    ];

    let rows = maze_lines.len();
    let cols = maze_lines[0].len();

    // Find S and E
    let mut start = None;
    let mut end = None;
    for (r, line) in maze_lines.iter().enumerate() {
        for (c, ch) in line.chars().enumerate() {
            match ch {
                'S' => start = Some((r, c)),
                'E' => end = Some((r, c)),
                _ => {}
            }
        }
    }

    let start = start.ok_or("Could not find 'S' in the maze.").unwrap();
    let end = end.ok_or("Could not find 'E' in the maze.").unwrap();

    // Dijkstra's algorithm
    const INF: usize = usize::MAX;
    let mut dist = vec![vec![vec![INF; 4]; cols]; rows];
    let start_dir = 0; // facing East
    dist[start.0][start.1][start_dir] = 0;

    let mut pq = BinaryHeap::new();
    pq.push(Reverse((0, start.0, start.1, start_dir)));

    let mut visited = HashSet::new();

    while let Some(Reverse((cost, r, c, d))) = pq.pop() {
        // If we've reached E, this is the minimal cost
        if (r, c) == end {
            return Some(cost);
        }

        if !visited.insert((r, c, d)) {
            continue;
        }

        let current_dist = dist[r][c][d];
        if cost > current_dist {
            continue;
        }

        // Explore neighbors:
        // 1) Move forward (cost + 1) if not a wall
        let (dr, dc) = directions[d];
        let nr = r as isize + dr;
        let nc = c as isize + dc;
        if nr >= 0 && nr < rows as isize && nc >= 0 && nc < cols as isize {
            let (nr, nc) = (nr as usize, nc as usize);
            if maze_lines[nr].chars().nth(nc).unwrap() != '#' {
                let new_cost = cost + 1;
                if new_cost < dist[nr][nc][d] {
                    dist[nr][nc][d] = new_cost;
                    pq.push(Reverse((new_cost, nr, nc, d)));
                }
            }
        }

        // 2) Turn left (cost + 1000)
        let left_dir = (d + 3) % 4;
        let new_cost = cost + 1000;
        if new_cost < dist[r][c][left_dir] {
            dist[r][c][left_dir] = new_cost;
            pq.push(Reverse((new_cost, r, c, left_dir)));
        }

        // 3) Turn right (cost + 1000)
        let right_dir = (d + 1) % 4;
        let new_cost = cost + 1000;
        if new_cost < dist[r][c][right_dir] {
            dist[r][c][right_dir] = new_cost;
            pq.push(Reverse((new_cost, r, c, right_dir)));
        }
    }

    None // If no path is found
}

fn solve_part2(maze_lines: &[String]) -> usize {
    let rows = maze_lines.len();
    let cols = maze_lines[0].len();

    // Directions: 0=East, 1=South, 2=West, 3=North
    let directions = [
        (0, 1),   // East
        (1, 0),   // South
        (0, -1),  // West
        (-1, 0),  // North
    ];

    let mut start = None;
    let mut end = None;

    for (r, line) in maze_lines.iter().enumerate() {
        for (c, ch) in line.chars().enumerate() {
            match ch {
                'S' => start = Some((r, c)),
                'E' => end = Some((r, c)),
                _ => {}
            }
        }
    }

    let start = start.expect("Could not find 'S' in the maze.");
    let end = end.expect("Could not find 'E' in the maze.");

    const INF: usize = usize::MAX;
    let mut dist = vec![vec![vec![INF; 4]; cols]; rows];
    dist[start.0][start.1][0] = 0;

    let mut pq = BinaryHeap::new();
    pq.push(Reverse((0, start.0, start.1, 0)));
    let mut visited = HashSet::new();

    while let Some(Reverse((cost, r, c, d))) = pq.pop() {
        if !visited.insert((r, c, d)) {
            continue;
        }

        let current_dist = dist[r][c][d];
        if cost > current_dist {
            continue;
        }

        let (dr, dc) = directions[d];
        let nr = r as isize + dr;
        let nc = c as isize + dc;
        if nr >= 0 && nr < rows as isize && nc >= 0 && nc < cols as isize {
            let (nr, nc) = (nr as usize, nc as usize);
            if maze_lines[nr].chars().nth(nc).unwrap() != '#' {
                let new_cost = cost + 1;
                if new_cost < dist[nr][nc][d] {
                    dist[nr][nc][d] = new_cost;
                    pq.push(Reverse((new_cost, nr, nc, d)));
                }
            }
        }

        for &turn in &[-1, 1] {
            let new_dir = (d as isize + turn).rem_euclid(4) as usize;
            let new_cost = cost + 1000;
            if new_cost < dist[r][c][new_dir] {
                dist[r][c][new_dir] = new_cost;
                pq.push(Reverse((new_cost, r, c, new_dir)));
            }
        }
    }

    let min_cost_end = (0..4).map(|d| dist[end.0][end.1][d]).min().unwrap();
    if min_cost_end == INF {
        return 0;
    }

    let mut on_best_path = vec![vec![false; cols]; rows];
    let mut queue = VecDeque::new();
    for d in 0..4 {
        if dist[end.0][end.1][d] == min_cost_end {
            queue.push_back((end.0, end.1, d));
        }
    }

    let mut visited_rev = HashSet::new();

    while let Some((r, c, d)) = queue.pop_front() {
        on_best_path[r][c] = true;
        let cost_here = dist[r][c][d];

        let (dr, dc) = directions[d];
        let prev_r = r as isize - dr;
        let prev_c = c as isize - dc;
        if prev_r >= 0 && prev_r < rows as isize && prev_c >= 0 && prev_c < cols as isize {
            let (prev_r, prev_c) = (prev_r as usize, prev_c as usize);
            if maze_lines[prev_r].chars().nth(prev_c).unwrap() != '#' {
                if cost_here >= 1 && dist[prev_r][prev_c][d] == cost_here - 1 {
                    if visited_rev.insert((prev_r, prev_c, d)) {
                        queue.push_back((prev_r, prev_c, d));
                    }
                }
            }
        }

        for &turn in &[-1, 1] {
            let prev_dir = (d as isize - turn).rem_euclid(4) as usize;
            let prev_cost = if cost_here >= 1000 { cost_here - 1000 } else { INF };
            if dist[r][c][prev_dir] == prev_cost {
                if visited_rev.insert((r, c, prev_dir)) {
                    queue.push_back((r, c, prev_dir));
                }
            }
        }
    }

    on_best_path.into_iter().flatten().filter(|&x| x).count()
}

fn main() {
    let input = std::fs::read_to_string("input.txt").expect("Failed to read input file");
    let maze_lines: Vec<String> = input.lines().map(|line| line.to_string()).collect();

    match solve_maze(&maze_lines) {
        Some(answer) => println!("Lowest possible score: {}", answer),
        None => println!("No path found!"),
    }

    let part2_result = solve_part2(&maze_lines);
    println!("Number of tiles on at least one best path: {}", part2_result);
}
